File: examples/latest/explore/flutter/routing_app/lib/TimeUtils.dart
import 'package:intl/intl.dart';
import 'package:here_sdk/core.dart';
import 'package:here_sdk/routing.dart';

class TimeUtils {
  /// Converts time in seconds to a formatted string in "HH:mm" format.
  ///
  /// @param sec the time in seconds to be converted.
  /// @return a string representing the time in "HH:mm" format.
  String formatTime(int sec) {
    // Calculate the number of hours from the given seconds
    int hours = sec ~/ 3600;

    // Calculate the number of minutes remaining after extracting hours
    int minutes = (sec % 3600) ~/ 60;

    // Format the hours and minutes into a string with "HH:mm" format
    return '$hours:$minutes min';
  }

  /// Converts length in meters to a formatted string in "km" format.
  ///
  /// @param meters the length in meters to be converted.
  /// @return a string representing the length in "km" format.
  String formatLength(int meters) {
    // Calculate the number of kilometers from the given meters
    int kilometers = meters ~/ 1000;

    // Calculate the remaining meters after extracting kilometers
    int remainingMeters = meters % 1000;

    // Format the kilometers and remaining meters into a string with "km" format
    return '$kilometers.$remainingMeters km';
  }

  /// Returns the ETA (as a string in ‘HH:mm’ format) in the current device’s timezone,
  /// derived from the estimatedTravelTimeInSeconds, which is sourced from the Route object.
  ///
  /// @param route Original route object from RoutingEngine.
  /// @return A string representing the ETA in "HH:mm" format.
  String getETAinDeviceTimeZone(Route route) {
    int estimatedTravelTimeInSeconds = route.duration.inSeconds;

    // Get the current date and time
    DateTime now = DateTime.now();

    // Add the estimated travel time (in seconds) to the current time
    DateTime eta = now.add(Duration(seconds: estimatedTravelTimeInSeconds));

    return DateFormat('hh:mm a').format(eta);
  }

  /// Calculates the estimated time of arrival (ETA) in the destination timezone for a given route.
  /// It is possible that the destination can be in a different timezone compared to the source.
  /// Therefore, we are also calculating the ETA in the destination timezone.
  ///
  /// @param route Original route object from RoutingEngine.
  /// @return A string representing the estimated time of arrival in the destination timezone, formatted as "HH:mm".
  String getETAinDestinationTimeZone(Route route) {
    DateTime destinationDate = getArrivalLocationTime(route).localTime;
    return DateFormat('hh:mm a').format(destinationDate);
  }

  /// Calculates the estimated time of arrival (ETA) in Coordinated Universal Time (UTC) for a given route.
  /// UTC (Coordinated Universal Time) is the primary time standard by which the world regulates clocks and time.
  ///
  /// @param route Original route object from RoutingEngine.
  /// @return A string representing the estimated time of arrival in UTC, formatted as "HH:mm".
  String getEstimatedTimeOfArrivalInUTC(Route route) {
    DateTime utcDate = getArrivalLocationTime(route).utcTime;

    // The UTC offset represents the difference in hours and minutes between a specific time zone and Coordinated Universal Time (UTC).
    // It indicates whether the local time is ahead (+) or behind (-) UTC.
    return DateFormat('hh:mm a').format(utcDate);
  }

  /// Returns the arrival time at the final location of the route.
  LocationTime getArrivalLocationTime(Route route) {
    int lastSectionIndex = route.sections.length - 1;

    // The lastSection contains cumulative duration values that increase sequentially.
    Section lastSection = route.sections[lastSectionIndex];
    return lastSection.arrivalLocationTime!;
  }
}

File: examples/latest/navigate/android/Java/heresdk-navigate-android-examples-4.23.4.0/Navigation/app/src/main/java/com/here/navigation/TimeUtils.java
package com.here.navigation;

import com.here.sdk.core.LocationTime;
import com.here.sdk.routing.Route;
import com.here.sdk.routing.Section;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

public class TimeUtils {

    /**
     * Converts time in seconds to a formatted string in "HH:mm" format.
     *
     * @param sec the time in seconds to be converted.
     * @return a string representing the time in "HH:mm" format.
     */
    public String formatTime(long sec) {
        // Calculate the number of hours from the given seconds
        int hours = (int) (sec / 3600);

        // Calculate the number of minutes remaining after extracting hours
        int minutes = (int) ((sec % 3600) / 60);

        // Format the hours and minutes into a string with "HH:mm" format
        return String.format(Locale.getDefault(), "%02d:%02d", hours, minutes);
    }

    /**
     * Converts length in meters to a formatted string in "km" format.
     *
     * @param meters the length in meters to be converted.
     * @return a string representing the length in "km" format.
     */
    public String formatLength(int meters) {
        // Calculate the number of kilometers from the given meters
        int kilometers = meters / 1000;

        // Calculate the remaining meters after extracting kilometers
        int remainingMeters = meters % 1000;

        // Format the kilometers and remaining meters into a string with "km" format
        return String.format(Locale.getDefault(), "%02d.%02d km", kilometers, remainingMeters);
    }

    /**
     * Returns the ETA (as a string in ‘HH:mm’ format) in the current device’s timezone, derived from the estimatedTravelTimeInSeconds, which is sourced from the Route object.
     *
     * @param remainingDuration remaining duration of trip in seconds.
     * @return A string representing the ETA in "HH:mm" format.
     */
    public String getETAinDeviceTimeZone(int remainingDuration) {

        // Get an instance of the Calendar class initialized with the current date and time
        Calendar calendar = Calendar.getInstance();

        // Set the calendar's time to the current date and time
        calendar.setTime(new Date());

        // Add the estimated travel time (in seconds) to the current time
        calendar.add(Calendar.SECOND, remainingDuration);

        return getFormattedDate(calendar.getTime());
    }


    /**
     * Formats the given date to a string representation using the device's default timezone.
     *
     * @param date The Date object to be formatted.
     * @return     A string representing the formatted time in the default timezone.
     */
    private String getFormattedDate(Date date) {
        // Create a DateFormat instance that formats the time in a short format (e.g., "HH:mm a").
        java.text.DateFormat dateFormat = SimpleDateFormat.getTimeInstance(java.text.DateFormat.SHORT);

        // Format the date using the configured DateFormat and return the result as a string.
        return dateFormat.format(date);
    }

}
